scheds/rust/scx_rusty/src/main.rs

```diff
@@ -543,68 +549,21 @@ impl<'a> Scheduler<'a> {
 
-        let numa_load_avg = lb_stats[0].load.load_avg();
-        let dom_load_avg = lb_stats[0].domains[0].load.load_avg();
-        info!(
-            "cpu={:7.2} bal={} numa_load_avg={:8.2} dom_load_avg={:8.2} task_err={} lb_data_err={} proc={:?}ms",
-            cpu_busy * 100.0,
-            bpf_stats[bpf_intf::stat_idx_RUSTY_STAT_LOAD_BALANCE as usize],
-            numa_load_avg, dom_load_avg,
-            bpf_stats[bpf_intf::stat_idx_RUSTY_STAT_TASK_GET_ERR as usize],
-            self.nr_lb_data_errors,
-            processing_dur.as_millis(),
-        );
-
-        let stat_pct = |value| value as f64 / total as f64 * 100.0;
-
-        info!(
-            "tot={:7} wsync_prev_idle={:5.2} wsync={:5.2}",
-            total,
-            stat_pct(wsync_prev_idle),
-            stat_pct(wsync),
-        );
-
-        info!(
-            "prev_idle={:5.2} greedy_idle={:5.2} pin={:5.2}",
-            stat_pct(prev_idle),
-            stat_pct(greedy_idle),
-            stat_pct(pinned),
-        );
-
-        info!(
-            "dir={:5.2} dir_greedy={:5.2} dir_greedy_far={:5.2}",
-            stat_pct(direct_dispatch),
-            stat_pct(direct_greedy),
-            stat_pct(direct_greedy_far),
-        );
-
-        info!(
-            "dsq={:5.2} greedy_local={:5.2} greedy_xnuma={:5.2}",
-            stat_pct(dsq),
-            stat_pct(greedy_local),
-            stat_pct(greedy_xnuma),
-        );
-
-        info!(
-            "kick_greedy={:5.2} rep={:5.2}",
-            stat_pct(kick_greedy),
-            stat_pct(repatriate),
-        );
-
-        info!(
-            "dl_clamped={:5.2} dl_preset={:5.2}",
-            stat_pct(dl_clamped),
-            stat_pct(dl_preset),
-        );
-
-        info!("slice_length={}us", self.tuner.slice_ns / 1000);
-        info!("direct_greedy_cpumask={}", self.tuner.direct_greedy_mask);
-        info!("  kick_greedy_cpumask={}", self.tuner.kick_greedy_mask);
+        self.metrics.task_errors.increment(stat(bpf_intf::stat_idx_RUSTY_STAT_TASK_GET_ERR));
+        self.metrics.lb_data_errors.increment(self.nr_lb_data_errors);
+        self.metrics.load_balance.increment(stat(bpf_intf::stat_idx_RUSTY_STAT_LOAD_BALANCE));
+        
+        self.metrics.slice_length.set(self.tuner.slice_ns as f64 / 1000.0);
 
+        // We need to dynamically create the  metrics for each node and domain 
+        // because we don't know how many there are at compile time. Metrics 
+        // will be cached and reused so this is not a performance issue.
         for node in lb_stats.iter() {
-            info!("{}", node);
+            histogram!("load_avg", "node" => node.id.to_string())
+                .record(node.load.load_avg() as f64);
             for dom in node.domains.iter() {
-                info!("{}", dom);
+                histogram!("load_avg", "node" => node.id.to_string(), "dom" => dom.id.to_string())
+                    .record(dom.load.load_avg() as f64);
```

Comment:
Let's not do this in this PR to keep the footprint small (we're already passing `lb_stats` to this function so I think what you have here is correct), but should we maybe move this into `load_balance.rs` instead? I think it makes sense for `main.rs` to print any of the statistics exported from BPF, but it might also be nice to let the load balancing subsystem track its own stats (and in general to allow schedulers to distribute how stats are collected).

Reply:
Yeah, this is an artifact of having to centralize metric printing in the `report` fn. I started moving some stuff out and this will be next. I wasn't sure how you'd feel about it but sounds like we are aligned :)